node/src/components/chain_synchronizer/operations.rs

```diff
@@ -948,34 +954,149 @@ where
+        .effect_builder
+        .query_global_state(QueryRequest::new(
+            *ctx.trusted_block_header().state_root_hash(),
+            Key::DeployApprovalsRootHash {
+                block_height: height,
+            },
+            vec![],
+        ))
+        .await
+        .map_err(FetchAndStoreBlockError::Engine)?
+    {
+        QueryResult::RootNotFound => {
+            error!("couldn't find the state root for the trusted header");
+            Err(FetchAndStoreBlockError::TrustedStateRootMissing)
+        }
+        QueryResult::DepthLimit { depth } => {
+            error!("reached depth limit: {:?}", depth);
+            Err(FetchAndStoreBlockError::Impossible(
+                "depth limit".to_string(),
+            ))
+        }
+        QueryResult::ValueNotFound(_err) => Ok(None),
+        QueryResult::CircularReference(err) => {
+            error!(
+                "circuluar reference for deploy approvals root hash: {:?}",
+                err
+            );
+            Err(FetchAndStoreBlockError::Impossible(err))
+        }
+        QueryResult::Success { value, proofs: _ } => match *value {
+            StoredValue::CLValue(cl_value) => {
+                let digest: Digest = cl_value
+                    .into_t()
+                    .map_err(FetchAndStoreBlockError::CLValue)?;
+                Ok(Some(digest))
+            }
+            stored_value => Err(
+                FetchAndStoreBlockError::WrongTypeUnderDeployApprovalsRootHash(Box::new(
+                    stored_value,
+                )),
+            ),
+        },
+    }
+}
+
 /// Downloads and stores a block with all its deploys.
 async fn fetch_and_store_block_with_deploys_by_hash<REv>(
     block_hash: BlockHash,
     ctx: &ChainSyncContext<'_, REv>,
-) -> Result<Box<BlockAndDeploys>, FetcherError<BlockAndDeploys>>
+) -> Result<Box<BlockAndDeploys>, FetchAndStoreBlockError>
 where
-    REv: From<NetworkInfoRequest> + From<FetcherRequest<BlockAndDeploys>> + From<StorageRequest>,
+    REv: From<NetworkInfoRequest>
+        + From<FetcherRequest<BlockAndDeploys>>
+        + From<StorageRequest>
+        + From<ContractRuntimeRequest>
+        + From<BlocklistAnnouncement>,
 {
     let start = Timestamp::now();
-    let fetched_block = fetch_retry_forever::<_, BlockAndDeploys>(ctx, block_hash).await?;
-    let res = match fetched_block {
-        FetchedData::FromStorage {
-            item: block_and_deploys,
-            ..
-        } => Ok(block_and_deploys),
-        FetchedData::FromPeer {
-            item: block_and_deploys,
-            ..
-        } => {
-            ctx.effect_builder
-                .put_block_and_deploys_to_storage(block_and_deploys.clone())
-                .await;
-            Ok(block_and_deploys)
-        }
-    };
-    ctx.metrics
-        .observe_fetch_block_and_deploys_duration_seconds(start);
-    res
+    'a: loop {
+        let fetched_block = fetch_retry_forever::<_, BlockAndDeploys>(ctx, block_hash)
```

Comment:
We're working on removing all `*_forever` instances from the code (cc @rafal-ch ). Can you make a note to fix that after that work is completed or follow the new approach? Rafał will know better whether we have the code to support that.

Reply:
I can probably just remove it here. What is the preferred approach?

Reply:
@SamuelSchlesinger I suggest leaving it as is for now as the preferred approach is still in development ([#3273](https://github.com/casper-network/casper-node/issues/3273), [#3274](https://github.com/casper-network/casper-node/issues/3274), [#3275](https://github.com/casper-network/casper-node/issues/3275)).

Depending on whether tickets above or your PR is done first, we'll then together adapt the code.